Incorrectly removing empty trailing closure parentheses in a curried call

## Problem

When calling `swift-format` on the following code

```swift
func perform() -> (() -> Void) -> Void {
  return { body in
    body()
  }
}

perform()() {
  print("Hello world")
}
```

It produces the following code (notice the `()` removed at the `perform` call)

```swift
func perform() -> (() -> Void) -> Void {
  return { body in
    body()
  }
}

perform() {
  print("hi")
}
```

This code doesn't compile and produces the following diagnostic.

```swift
File.swift:7:11: error: extra trailing closure passed in call
 5 | }
 6 | 
 7 | perform() {
   |           `- error: extra trailing closure passed in call
 8 |   print("Hello world")
 9 | }
```

The above example is also easily reproducible with implicitly-curried instance methods as in the following example.

```swift
struct Executor {
  func execute(body: () throws -> Void) rethrows {
    body()
  }
}

let executor = Executor()

Executor.execute(executor)() { }
```
## Proposed Solution

`swift-format` should only remove empty function arguments `()` before a trailing closure when the `FunctionCallExpr`'s callee is not a `FunctionCallExpr` nor a `SubscriptCallExpr` to avoid incorrectly associating the trailing closure with the callee.

--- a/Sources/SwiftFormat/Rules/NoEmptyTrailingClosureParentheses.swift
+++ b/Sources/SwiftFormat/Rules/NoEmptyTrailingClosureParentheses.swift
@@ -38,6 +38,15 @@ public final class NoEmptyTrailingClosureParentheses: SyntaxFormatRule {
       return super.visit(node)
     }
 
+    // Keep the empty parentheses when in a curried call to avoid the trailing closure
+    // getting associated with the called call expression.
+    guard
+      !node.calledExpression.is(FunctionCallExprSyntax.self)
+        && !node.calledExpression.is(SubscriptCallExprSyntax.self)
+    else {
+      return super.visit(node)
+    }
+
     diagnose(.removeEmptyTrailingParentheses(name: "\(name.trimmedDescription)"), on: leftParen)
 
     // Need to visit `calledExpression` before creating a new node so that the location data (column

--- a/Tests/SwiftFormatTests/Rules/NoEmptyTrailingClosureParenthesesTests.swift
+++ b/Tests/SwiftFormatTests/Rules/NoEmptyTrailingClosureParenthesesTests.swift
@@ -46,7 +46,7 @@ final class NoEmptyTrailingClosureParenthesesTests: LintOrFormatRuleTestCase {
             greetEnthusiastically8️⃣() { "Willis" }
           }
         }
-        foo(bar🔟() { baz })9️⃣() { blah }
+        foo(bar9️⃣() { baz }) { blah }
         """,
       expected: """
         func greetEnthusiastically(_ nameProvider: () -> String) {
@@ -89,8 +89,7 @@ final class NoEmptyTrailingClosureParenthesesTests: LintOrFormatRuleTestCase {
         FindingSpec("6️⃣", message: "remove the empty parentheses following 'async'"),
         FindingSpec("7️⃣", message: "remove the empty parentheses following 'greetEnthusiastically'"),
         FindingSpec("8️⃣", message: "remove the empty parentheses following 'greetEnthusiastically'"),
-        FindingSpec("9️⃣", message: "remove the empty parentheses following ')'"),
-        FindingSpec("🔟", message: "remove the empty parentheses following 'bar'"),
+        FindingSpec("9️⃣", message: "remove the empty parentheses following 'bar'"),
       ]
     )
   }
@@ -115,6 +114,23 @@ final class NoEmptyTrailingClosureParenthesesTests: LintOrFormatRuleTestCase {
         greetEnthusiastically(
           // oldArg: x
         ) { "John" }
+        """,
+      findings: []
+    )
+  }
+
+  func testDoNotRemoveParensInCurriedCalls() {
+    assertFormatting(
+      NoEmptyTrailingClosureParentheses.self,
+      input: """
+        perform()() { foo }
+        Executor.execute(executor)() { bar }
+        withSubscript[baz]() { blah }
+        """,
+      expected: """
+        perform()() { foo }
+        Executor.execute(executor)() { bar }
+        withSubscript[baz]() { blah }
         """,
       findings: []
     )

